pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,303 @@
+	ips []string, ruleIdx int) *model.Group {
+	ipSetGroup := model.Group{}
+
+	ipSetGroupID := fmt.Sprintf("%s_ipset", *ruleModel.Id)
+	ipSetGroup.Id = &ipSetGroupID
+	ipSetGroupName := fmt.Sprintf("%s-ipset", *ruleModel.DisplayName)
+	ipSetGroup.DisplayName = &ipSetGroupName
+	peerTags := service.BuildPeerTags(obj, &rule.Destinations, ruleIdx)
+	ipSetGroup.Tags = peerTags
+
+	addresses := data.NewListValue()
+	for _, ip := range ips {
+		addresses.Add(data.NewStringValue(ip))
+	}
+
+	blockExpression := data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"resource_type": data.NewStringValue("IPAddressExpression"),
+			"ip_addresses":  addresses,
+		},
+	)
+	ipSetGroup.Expression = append(ipSetGroup.Expression, blockExpression)
+	return &ipSetGroup
+}
+
+// Different direction rule decides different target of the traffic, we should carefully get
+// the destination pod selector and namespaces. Named port only cares about the destination
+// pod selector, so we use policy's AppliedTo or rule's AppliedTo when "IN" direction and
+// rule's DestinationSelector when "OUT" direction.
+func (service *SecurityPolicyService) getPodSelectors(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule) ([]client.ListOptions, error) {
+	// Port means the target of traffic, so we should select the pod by rule direction,
+	// as for IN direction, we judge by rule's or policy's AppliedTo,
+	// as for OUT direction, then by rule's destinations.
+	// LabelSelect may return multiple namespaces
+	var selectorsUltimate []client.ListOptions
+	ruleDirection, err := getRuleDirection(rule)
+	if err != nil {
+		return nil, err
+	}
+
+	if ruleDirection == "IN" {
+		if len(obj.Spec.AppliedTo) > 0 {
+			for _, target := range obj.Spec.AppliedTo {
+				selector := client.ListOptions{}
+				if target.PodSelector != nil {
+					label, err := meta1.LabelSelectorAsSelector(target.PodSelector)
+					if err != nil {
+						return nil, err
+					}
+					selector.LabelSelector = label
+					selector.Namespace = obj.Namespace
+					selectorsUltimate = append(selectorsUltimate, selector)
+				}
+			}
+		} else if len(rule.AppliedTo) > 0 {
+			for _, target := range rule.AppliedTo {
+				// We only consider named port for PodSelector, not VMSelector
+				selector := client.ListOptions{}
+				if target.PodSelector != nil {
+					label, err := meta1.LabelSelectorAsSelector(target.PodSelector)
+					if err != nil {
+						return nil, err
+					}
+					selector.LabelSelector = label
+					selector.Namespace = obj.Namespace
+					selectorsUltimate = append(selectorsUltimate, selector)
+				}
+			}
+		}
+	} else if ruleDirection == "OUT" {
+		if len(rule.Destinations) > 0 {
+			var labelSelectors []client.ListOptions
+			var namespaceSelectors []client.ListOptions
+			for _, target := range rule.Destinations {
+				labelSelector := client.ListOptions{}
+				namespaceSelector := client.ListOptions{}
+				if target.PodSelector != nil {
```

Comment:
Do we need to handle target.PodSelector == nil?

Reply:
The logic considers this.